x/gamm/pool-models/stableswap/amm.go

```diff
@@ -291,22 +292,26 @@ func solveCFMMBinarySearch(constantFunction func(osmomath.BigDec, osmomath.BigDe
 	return func(xReserve, yReserve, yIn osmomath.BigDec) osmomath.BigDec {
 		if !xReserve.IsPositive() || !yReserve.IsPositive() || !yIn.IsPositive() {
 			panic("invalid input: reserves and input must be positive")
+		} else if yIn.GTE(yReserve) {
+			panic("cannot input more than pool reserves")
 		}
 		k := constantFunction(xReserve, yReserve)
 		yf := yReserve.Add(yIn)
 		x_low_est := osmomath.ZeroDec()
 		x_high_est := xReserve
-		x_est := (x_high_est.Add(x_low_est)).Quo(twodec)
-		cur_k := constantFunction(x_est, yf)
-		for i := 0; !approxDecEqual(cur_k, k, k_threshold) && i < 256; i++ {
-			if cur_k.GT(k) {
-				x_high_est = x_est
-			} else if cur_k.LT(k) {
-				x_low_est = x_est
-			}
-			x_est = (x_high_est.Add(x_low_est)).Quo(twodec)
-			cur_k = constantFunction(x_est, yf)
+		max_iterations := 256
+		errTolerance := osmoutils.ErrTolerance{AdditiveTolerance: sdk.OneInt(), MultiplicativeTolerance: sdk.Dec{}}
+
+		// create single-input CFMM to pass into binary search
+		calc_x_est := func(xEst osmomath.BigDec) (osmomath.BigDec, error) {
+			return cfmmConstant(xEst, yf), nil
```

Comment:
If we don't want to replace the `constantFunction` with `cfmmConstant` above, do we want to do the opposite here?

Reply:
https://github.com/osmosis-labs/osmosis/pull/2816#discussion_r978020447